lib/srv/app/server.go

```diff
@@ -956,12 +937,31 @@ func (s *Server) getApp(ctx context.Context, publicAddr string) (types.Applicati
 
 	for _, a := range s.getApps() {
 		if publicAddr == a.GetPublicAddr() {
-			return a, nil
+			return s.appWithUpdatedLabels(a), nil
 		}
 	}
 	return nil, trace.NotFound("no application at %v found", publicAddr)
 }
 
+func (s *Server) appWithUpdatedLabels(app types.Application) *types.AppV3 {
```

Comment:
You should either take the `RLock` here, bring back `getDynamicLabels`, or clearly document that the caller must hold the `RLock` before calling this function.

Reply:
I'll document it -- I removed `getDynamicLabels` because it was only used in one place and the two callers of `appWithUpdatedLabels` have different semantics about locking.